lib/Runtime/Types/TypePropertyCache.cpp

```diff
@@ -448,10 +448,32 @@ namespace Js
             element.Clear();
     }
 
+    bool TypePropertyCache::ClearIfPropertyIsMissing(const PropertyId id)
+    {
+        TypePropertyCacheElement &element = elements[ElementIndex(id)];
+        if (element.Id() == id && element.IsMissing())
+        {
+            element.Clear();
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+
     void TypePropertyCache::Clear(const PropertyId id)
     {
         TypePropertyCacheElement &element = elements[ElementIndex(id)];
         if(element.Id() == id)
             element.Clear();
     }
+
+#if DBG
+    bool TypePropertyCache::PropertyIsMissing(const PropertyId id) const
```

Comment:
There's a risk of a debug-only method not getting updated when the corresponding "real" method gets changed, and leading to confusion later on. Could we maybe have ClearIfPropertyIsMissing call PropertyIsMissing to avoid duplication, and set PropertyIsMissing to either public or private in the header depending on the build type? (Alternately, it's just two lines, so feel free to leave as-is.)

Reply:
Yes, I agree that this is a danger in the general case.  However, since this method is called in an assertion (ThreadContext.cpp:4087, in InvalidateMissingPropertyProtoTypePropertyCaches), we should have sufficient test coverage in, e.g., VSTS builds that the method implementation won't rot.  That said, I don't have strong feelings about this, so I can remove the #if DBG/#endif if people would prefer.  (Unfortunately, I think this particular bit of the code is old enough that I no longer entirely remember my motivations for marking it DBG-only.)

Reply:
>implementation won't rot

Well, it won't rot in _one_ of the two possible directions: it could still rot in such a way that it returns true when it shouldn't, and the assertion would pass and hide a real bug. Anyhow, I agree that the risk is small enough that we shouldn't worry about it.

---
In reply to: [173555571](https://github.com/Microsoft/ChakraCore/pull/4780#discussion_r173555571) [](ancestors = 173555571)

Reply:
I could have ClearIfPropertyIsMissing call PropertyIsMissing, but only at the cost of either duplicating half of the body of PropertyIsMissing (admittedly only a single line) or of redefining the interface of both methods in terms of TypePropertyCacheElement instead of PropertyId.  The latter in particular seems an unfortunate instance of information leakage: although the TypePropertyCacheElement type is visible to clients of TypePropertyCache, it's not something that clients of the cache currently have to use directly.